### PR TITLE
C4C180-UniqueExtras

### PR DESCRIPTION
**Pull request by rekowp17**
I created a stored procedure to return randomized addons; updated controller method GetExtras; Created test to ensure distinct extras sets

### CODE REVIEW

File: C4CCoreApp/C4CAndroid/AddonsActivity.cs [hunk 1]
```diff
@@ -153,7 +153,6 @@ private void CheckboxClick(object sender, EventArgs e)
             using (var httpClient = new HttpClient(new NativeMessageHandler()))
             {
                 // The URL of the web server, currently locally hosted.
-
                 string url = "http://10.0.2.2:5000/api/Data/GetExtras";
```
Comment (murphyjp) [inline C4CCoreApp/C4CAndroid/AddonsActivity.cs]: Not specific to this PR, but for a future feature it'd be nice to have a static variable somewhere that holds this url so if/when it changes, it only needs to be updated in one place. Likewise a shared Connection that can handle its own open/close so we don't have to open/close the connection every time it's used. Perhaps the GetDBConnection could handle opening, if not open, and disposal of that object can handle closing. Just an idea.

File: C4CCoreApp/C4CCoreApp/Controllers/DataController.cs [hunk 2]
```diff
@@ -51,18 +51,25 @@ public HttpResponseMessage GetCancers()
...
 
-            // Select three random addons to display.
-            Random randomGen = new Random();
-            Item[] randomAddons = new Item[3];
-            int randomCeiling = allAddons.Count - 1;
-            for (int i = 0; i < 3; i++)
+            try
             {
-                randomAddons[i] = allAddons[randomGen.Next(0, randomCeiling)];
+                conn.Open();
```
Comment (kubowitsche) [inline C4CCoreApp/C4CCoreApp/Controllers/DataController.cs]: could you put the conn.open/conn.close try/catch into Item.GetAvailableAddons? Just to keep consistent with the style that we agreed on yesterday

File: C4CCoreApp/C4CCoreTests/DatabaseConnectorTests.cs [hunk 3]
```diff
@@ -284,18 +284,73 @@ public void TestGetAvailableBoxTypes()
...
+
+            List<List<Item>> keep_addons = new List<List<Item>>();
+            List<Item> addons = new List<Item>();
+            int diffMark = 0;
+
+            for (int i = 0; i < repetitions; ++i)
+            {
+                try
+                {
+                    conn.Open();
```
Comment (kubowitsche) [inline C4CCoreApp/C4CCoreTests/DatabaseConnectorTests.cs]: again, should be in Item.Get AvailableAddons

File: Database/gAvailableAddonsV2.sql [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    INSERT INTO r2 SELECT * FROM Ranked_Addons;
+    
+    SELECT r1.ItemID, r1.TypeID, r1.IsAddon, r1.Options, r1.QuantityLeft, r1.Name, r1.Price, r1.ItemDescription, r1.IntentionID
+	FROM Ranked_Addons AS r1
+	LEFT JOIN r2
+		ON r1.TypeID = r2.TypeID
+        AND r1.Rank < r2.Rank
+	WHERE r2.TypeID IS NULL
+    ORDER BY r1.Rank DESC
+    LIMIT 3;
```
Comment (kubowitsche) [inline Database/gAvailableAddonsV2.sql]: So how exactly is this grabbing unique addons?  Maybe I'm not understanding the joins correctly, but would you mind explaining your process?

**Pull request merged**